compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -155,9 +169,54 @@ impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
                     .map(|of_trait| path_has_local_parent(of_trait.path, cx, parent, parent_parent))
                     .unwrap_or(false);
 
-                // If none of them have a local parent (LOGICAL NOR) this means that
-                // this impl definition is a non-local definition and so we lint on it.
-                if !(self_ty_has_local_parent || of_trait_has_local_parent) {
+                // Detecting if the impl definition is leaking outside of it's defining scope.
+                //
+                // Rule: for each impl, instantiate all local types with inference vars and
+                // then assemble candidates for that goal, if there are more than 1 (non-private
+                // impls), it does not leak.
+                //
+                // https://github.com/rust-lang/rust/issues/121621#issuecomment-1976826895
+                let impl_is_not_leaky = cx
```

Comment:
please move this into a separate function

Reply:
Moved into a separate function.